File: android/app/src/test/java/com/hawksnest/core/logic/ClipExportTest.kt
package com.hawksnest.core.logic

import java.util.Calendar
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The parity contract for clip export. `src/lib/__tests__/clipExport.test.ts` carries the *same
 * cases with the same numbers* — the two files are meant to be diffable. If you change an
 * assertion here, change it there, or the platforms have silently forked.
 */
class ClipExportTest {
    private val now = 1_800_000_000_000L
    private val day = 86_400_000L

    /** A 3-day retention window padded past now, as the timeline actually supplies it. */
    private val window = TimeWindow(now - 3 * day, now + 1_800_000L)
    private val b = exportBounds(window, now)

    private fun span(startMs: Long, endMs: Long, playable: Boolean = true) =
        FootageSpan(startMs, endMs, playable)

    // ---- exportBounds ----

    @Test
    fun `exportBounds stops one segment short of now`() {
        assertEquals(now - EXPORT_TAIL_LAG_MS, exportBounds(window, now).endMs)
    }

    @Test
    fun `exportBounds keeps the retention floor and never inverts`() {
        assertEquals(window.startMs, exportBounds(window, now).startMs)
        val tiny = TimeWindow(now, now)
        assertEquals(now, exportBounds(tiny, now).endMs)
    }

    @Test
    fun `exportBounds honours a window ending before the tail lag would`() {
        val early = TimeWindow(now - day, now - 10 * 60_000L)
        assertEquals(early.endMs, exportBounds(early, now).endMs)
    }

    // ---- defaultSelection ----

    @Test
    fun `defaultSelection opens centred on the playhead`() {
        val head = now - day
        assertEquals(
            ClipSelection(head - DEFAULT_HALF_MS, head + DEFAULT_HALF_MS),
            defaultSelection(head, b),
        )
    }

    @Test
    fun `defaultSelection slides inside the bounds rather than shrinking at an edge`() {
        val sel = defaultSelection(b.endMs, b)
        assertEquals(b.endMs, sel.endMs)
        assertEquals(2 * DEFAULT_HALF_MS, selectionDurationMs(sel))
    }

    // ---- clampSelection ----

    @Test
    fun `clampSelection grows a too-short range to the minimum`() {
        val sel = clampSelection(ClipSelection(now - day, now - day + 500L), b)
        assertEquals(MIN_CLIP_MS, selectionDurationMs(sel))
    }

    @Test
    fun `clampSelection trims a too-long range to the maximum`() {
        val sel = clampSelection(ClipSelection(now - day, now - day + 3 * MAX_CLIP_MS), b)
        assertEquals(MAX_CLIP_MS, selectionDurationMs(sel))
    }

    @Test
    fun `clampSelection moves the edge the user is NOT holding`() {
        val start = now - day
        val over = ClipSelection(start, start + 3 * MAX_CLIP_MS)
        assertEquals(
            ClipSelection(start, start + MAX_CLIP_MS),
            clampSelection(over, b, ClipEdge.START),
        )
        assertEquals(
            ClipSelection(over.endMs - MAX_CLIP_MS, over.endMs),
            clampSelection(over, b, ClipEdge.END),
        )
    }

    @Test
    fun `clampSelection never returns an inverted range`() {
        val sel = clampSelection(ClipSelection(now - day, now - 2 * day), b)
        assertTrue(sel.endMs >= sel.startMs)
    }

    @Test
    fun `clampSelection collapses when the bounds are narrower than a minimum clip`() {
        val narrow = TimeWindow(now, now + 1_000L)
        assertEquals(narrow.startMs, clampSelection(ClipSelection(now - day, now + day), narrow).startMs)
        assertEquals(narrow.endMs, clampSelection(ClipSelection(now - day, now + day), narrow).endMs)
    }

    // ---- setEdge ----

    @Test
    fun `setEdge keeps the moved edge exactly where it was put`() {
        val sel = ClipSelection(now - day, now - day + 60_000L)
        val target = now - day + 20_000L
        assertEquals(target, setEdge(sel, ClipEdge.START, target, b).startMs)
        assertEquals(target, setEdge(sel, ClipEdge.END, target, b).endMs)
    }

    @Test
    fun `setEdge pins rather than swaps when dragged past the other edge`() {
        val sel = ClipSelection(now - day, now - day + 60_000L)
        val dragged = setEdge(sel, ClipEdge.START, sel.endMs + 60_000L, b)
        assertTrue(dragged.startMs < dragged.endMs)
        assertEquals(MIN_CLIP_MS, selectionDurationMs(dragged))
    }

    @Test
    fun `setEdge cannot place an edge past the export ceiling`() {
        val sel = ClipSelection(now - day, now - day + 60_000L)
        assertEquals(b.endMs, setEdge(sel, ClipEdge.END, now + day, b).endMs)
    }

    @Test
    fun `setEdge cannot place an edge before retention`() {
        val sel = ClipSelection(now - day, now - day + 60_000L)
        assertEquals(b.startMs, setEdge(sel, ClipEdge.START, b.startMs - day, b).startMs)
    }

    // ---- nudge ----

    @Test
    fun `nudge shifts one edge and leaves the other alone`() {
        val sel = ClipSelection(now - day, now - day + 60_000L)
        val out = nudge(sel, ClipEdge.END, 15_000L, b)
        assertEquals(sel.endMs + 15_000L, out.endMs)
        assertEquals(sel.startMs, out.startMs)
    }

    @Test
    fun `nudge refuses to go below the minimum duration`() {
        val sel = ClipSelection(now - day, now - day + MIN_CLIP_MS)
        assertEquals(MIN_CLIP_MS, selectionDurationMs(nudge(sel, ClipEdge.END, -60_000L, b)))
    }

    // ---- pickHandle ----

    @Test
    fun `pickHandle returns null when near neither handle`() {
        assertNull(pickHandle(500f, 100f, 200f))
    }

    @Test
    fun `pickHandle grabs whichever handle is within slop`() {
        assertEquals(ClipEdge.START, pickHandle(100f + HANDLE_HIT_SLOP_PX - 1f, 100f, 400f))
        assertEquals(ClipEdge.END, pickHandle(400f - HANDLE_HIT_SLOP_PX + 1f, 100f, 400f))
    }

    @Test
    fun `pickHandle prefers the nearer handle and breaks a tie towards end`() {
        assertEquals(ClipEdge.START, pickHandle(105f, 100f, 130f))
        assertEquals(ClipEdge.END, pickHandle(115f, 100f, 130f))
        assertEquals(ClipEdge.END, pickHandle(100f, 100f, 100f))
    }

    // ---- coverage ----

    private val covSel = ClipSelection(now - day, now - day + 60_000L)

    @Test
    fun `coverage reports unknown - not none - when there is no lane data`() {
        // The lane is absent while the fetch is in flight and on sources that don't provide one.
        // Reporting NONE there would permanently disable export on a transient failure.
        assertEquals(ClipCoverage.UNKNOWN, coverage(covSel, emptyList()))
    }

    @Test
    fun `coverage reports full when a span encloses the selection`() {
        val spans = listOf(span(covSel.startMs - 60_000L, covSel.endMs + 60_000L))
        assertEquals(ClipCoverage.FULL, coverage(covSel, spans))
    }

    @Test
    fun `coverage reports none when no span overlaps`() {
        val spans = listOf(span(now - 2 * day, now - 2 * day + 60_000L))
        assertEquals(ClipCoverage.NONE, coverage(covSel, spans))
    }

    @Test
    fun `coverage reports partial when the selection straddles a gap`() {
        val spans =
            listOf(
                span(covSel.startMs - 60_000L, covSel.startMs + 15_000L),
                span(covSel.endMs - 15_000L, covSel.endMs + 60_000L),
            )
        assertEquals(ClipCoverage.PARTIAL, coverage(covSel, spans))
    }

    @Test
    fun `coverage does not count footage that cannot be decoded`() {
        val spans = listOf(span(covSel.startMs, covSel.endMs, playable = false))
        assertEquals(ClipCoverage.NONE, coverage(covSel, spans))
    }

    // ---- selectionProblem ----

    private val covered = listOf(span(now - 3 * day, now))

    @Test
    fun `selectionProblem passes a well-formed covered selection`() {
        assertNull(selectionProblem(ClipSelection(now - day, now - day + 60_000L), covered))
    }

    @Test
    fun `selectionProblem rejects durations outside the limits`() {
        assertEquals(
            ClipProblem.TOO_SHORT,
            selectionProblem(ClipSelection(now - day, now - day + 1_000L), covered),
        )
        assertEquals(
            ClipProblem.TOO_LONG,
            selectionProblem(ClipSelection(now - day, now - day + MAX_CLIP_MS + 1_000L), covered),
        )
    }

    @Test
    fun `selectionProblem rejects a range with no footage behind it`() {
        assertNull(selectionProblem(ClipSelection(now - day, now - day + 60_000L), emptyList()))
        assertEquals(
            ClipProblem.NO_FOOTAGE,
            selectionProblem(
                ClipSelection(now - day, now - day + 60_000L),
                listOf(span(now - 2 * day, now - 2 * day + 1_000L)),
            ),
        )
    }

    @Test
    fun `selectionProblem allows a partially covered range`() {
        val sel = ClipSelection(now - day, now - day + 60_000L)
        assertNull(selectionProblem(sel, listOf(span(sel.startMs, sel.startMs + 20_000L))))
    }

    // ---- clipRangeSeconds ----

    @Test
    fun `clipRangeSeconds rounds outwards`() {
        assertEquals(ClipRange(1_000L, 2_001L), clipRangeSeconds(ClipSelection(1_000_400L, 2_000_100L)))
    }

    @Test
    fun `clipRangeSeconds leaves whole seconds untouched`() {
        assertEquals(ClipRange(1_000L, 2_000L), clipRangeSeconds(ClipSelection(1_000_000L, 2_000_000L)))
    }

    // ---- clipFileName ----

    private fun localMs(y: Int, mon: Int, d: Int, h: Int, min: Int, s: Int): Long {
        val cal = Calendar.getInstance()
        cal.clear()
        cal.set(y, mon, d, h, min, s)
        return cal.timeInMillis
    }

    @Test
    fun `clipFileName stamps the clip's LOCAL start time and its length`() {
        val start = localMs(2026, 7, 9, 14, 3, 5)
        assertEquals(
            "big_room-2026-08-09-14-03-05-50s.mp4",
            clipFileName("big_room", ClipSelection(start, start + 50_000L)),
        )
    }

    @Test
    fun `clipFileName strips anything that has no business in a filename`() {
        val start = localMs(2026, 7, 9, 1, 2, 3)
        assertEquals(
            "front-door-2026-08-09-01-02-03-10s.mp4",
            clipFileName("../front door", ClipSelection(start, start + 10_000L)),
        )
    }
}

File: src/lib/__tests__/clipExport.test.ts
import { describe, expect, it } from "vitest";
import type { FootageSpan } from "../ringFootage";
import {
  DEFAULT_HALF_MS,
  EXPORT_TAIL_LAG_MS,
  HANDLE_HIT_SLOP_PX,
  MAX_CLIP_MS,
  MIN_CLIP_MS,
  clampSelection,
  clipFileName,
  clipRangeSeconds,
  coverage,
  defaultSelection,
  exportBounds,
  nudge,
  pickHandle,
  selectionDurationMs,
  selectionProblem,
  setEdge,
} from "../clipExport";

/**
 * The parity contract for clip export. `core/logic/ClipExportTest.kt` carries the *same cases with
 * the same numbers* — the two files are meant to be diffable. If you change an assertion here,
 * change it there, or the platforms have silently forked.
 */

const NOW = 1_800_000_000_000;
const DAY = 86_400_000;
/** A 3-day retention window padded past now, as the timeline actually supplies it. */
const WINDOW = { startMs: NOW - 3 * DAY, endMs: NOW + 1_800_000 };
const B = exportBounds(WINDOW, NOW);

const span = (startMs: number, endMs: number, playable = true): FootageSpan => ({
  startMs,
  endMs,
  playable,
});

describe("exportBounds", () => {
  it("stops one segment short of now, so the segment still being written is never requested", () => {
    expect(exportBounds(WINDOW, NOW).endMs).toBe(NOW - EXPORT_TAIL_LAG_MS);
  });

  it("keeps the retention floor and never inverts when the window is degenerate", () => {
    expect(exportBounds(WINDOW, NOW).startMs).toBe(WINDOW.startMs);
    const tiny = { startMs: NOW, endMs: NOW };
    expect(exportBounds(tiny, NOW).endMs).toBe(NOW);
  });

  it("honours a window that ends before the tail lag would", () => {
    const early = { startMs: NOW - DAY, endMs: NOW - 10 * 60_000 };
    expect(exportBounds(early, NOW).endMs).toBe(early.endMs);
  });
});

describe("defaultSelection", () => {
  it("opens centred on the playhead", () => {
    const head = NOW - DAY;
    expect(defaultSelection(head, B)).toEqual({
      startMs: head - DEFAULT_HALF_MS,
      endMs: head + DEFAULT_HALF_MS,
    });
  });

  it("slides inside the bounds rather than shrinking when the playhead is at an edge", () => {
    const sel = defaultSelection(B.endMs, B);
    expect(sel.endMs).toBe(B.endMs);
    expect(selectionDurationMs(sel)).toBe(2 * DEFAULT_HALF_MS);
  });
});

describe("clampSelection", () => {
  it("grows a too-short range to the minimum", () => {
    const sel = clampSelection({ startMs: NOW - DAY, endMs: NOW - DAY + 500 }, B);
    expect(selectionDurationMs(sel)).toBe(MIN_CLIP_MS);
  });

  it("trims a too-long range to the maximum", () => {
    const sel = clampSelection({ startMs: NOW - DAY, endMs: NOW - DAY + 3 * MAX_CLIP_MS }, B);
    expect(selectionDurationMs(sel)).toBe(MAX_CLIP_MS);
  });

  it("moves the edge the user is NOT holding", () => {
    const start = NOW - DAY;
    const over = { startMs: start, endMs: start + 3 * MAX_CLIP_MS };
    // Anchored on start: the end yields.
    expect(clampSelection(over, B, "start")).toEqual({ startMs: start, endMs: start + MAX_CLIP_MS });
    // Anchored on end: the start yields instead.
    expect(clampSelection(over, B, "end")).toEqual({
      startMs: over.endMs - MAX_CLIP_MS,
      endMs: over.endMs,
    });
  });

  it("never returns an inverted range", () => {
    const sel = clampSelection({ startMs: NOW - DAY, endMs: NOW - 2 * DAY }, B);
    expect(sel.endMs).toBeGreaterThanOrEqual(sel.startMs);
  });

  it("collapses to the whole range when the bounds are narrower than a minimum clip", () => {
    const narrow = { startMs: NOW, endMs: NOW + 1_000 };
    expect(clampSelection({ startMs: NOW - DAY, endMs: NOW + DAY }, narrow)).toEqual(narrow);
  });
});

describe("setEdge", () => {
  it("keeps the moved edge exactly where it was put", () => {
    const sel = { startMs: NOW - DAY, endMs: NOW - DAY + 60_000 };
    const target = NOW - DAY + 20_000;
    expect(setEdge(sel, "start", target, B).startMs).toBe(target);
    expect(setEdge(sel, "end", target, B).endMs).toBe(target);
  });

  it("pins rather than swaps when an edge is dragged past the other", () => {
    const sel = { startMs: NOW - DAY, endMs: NOW - DAY + 60_000 };
    const dragged = setEdge(sel, "start", sel.endMs + 60_000, B);
    expect(dragged.startMs).toBeLessThan(dragged.endMs);
    expect(selectionDurationMs(dragged)).toBe(MIN_CLIP_MS);
  });

  it("cannot place an edge past the export ceiling", () => {
    const sel = { startMs: NOW - DAY, endMs: NOW - DAY + 60_000 };
    expect(setEdge(sel, "end", NOW + DAY, B).endMs).toBe(B.endMs);
  });

  it("cannot place an edge before retention", () => {
    const sel = { startMs: NOW - DAY, endMs: NOW - DAY + 60_000 };
    expect(setEdge(sel, "start", B.startMs - DAY, B).startMs).toBe(B.startMs);
  });
});

describe("nudge", () => {
  it("shifts one edge by the delta and leaves the other alone", () => {
    const sel = { startMs: NOW - DAY, endMs: NOW - DAY + 60_000 };
    const out = nudge(sel, "end", 15_000, B);
    expect(out.endMs).toBe(sel.endMs + 15_000);
    expect(out.startMs).toBe(sel.startMs);
  });

  it("refuses to nudge below the minimum duration", () => {
    const sel = { startMs: NOW - DAY, endMs: NOW - DAY + MIN_CLIP_MS };
    expect(selectionDurationMs(nudge(sel, "end", -60_000, B))).toBe(MIN_CLIP_MS);
  });
});

describe("pickHandle", () => {
  it("returns null when the press is near neither handle", () => {
    expect(pickHandle(500, 100, 200)).toBeNull();
  });

  it("grabs whichever handle is within slop", () => {
    expect(pickHandle(100 + HANDLE_HIT_SLOP_PX - 1, 100, 400)).toBe("start");
    expect(pickHandle(400 - HANDLE_HIT_SLOP_PX + 1, 100, 400)).toBe("end");
  });

  it("prefers the nearer handle, and breaks an exact tie towards end", () => {
    expect(pickHandle(105, 100, 130)).toBe("start");
    expect(pickHandle(115, 100, 130)).toBe("end");
    expect(pickHandle(100, 100, 100)).toBe("end");
  });
});

describe("coverage", () => {
  const sel = { startMs: NOW - DAY, endMs: NOW - DAY + 60_000 };

  it("reports unknown — not none — when there is no lane data at all", () => {
    // The lane is absent while the fetch is in flight and on sources that don't provide one.
    // Reporting "none" there would permanently disable export on a transient failure.
    expect(coverage(sel, [])).toBe("unknown");
  });

  it("reports full when a span encloses the selection", () => {
    expect(coverage(sel, [span(sel.startMs - 60_000, sel.endMs + 60_000)])).toBe("full");
  });

  it("reports none when no span overlaps", () => {
    expect(coverage(sel, [span(NOW - 2 * DAY, NOW - 2 * DAY + 60_000)])).toBe("none");
  });

  it("reports partial when the selection straddles a gap", () => {
    const spans = [
      span(sel.startMs - 60_000, sel.startMs + 15_000),
      span(sel.endMs - 15_000, sel.endMs + 60_000),
    ];
    expect(coverage(sel, spans)).toBe("partial");
  });

  it("does not count footage that exists but cannot be decoded", () => {
    expect(coverage(sel, [span(sel.startMs, sel.endMs, false)])).toBe("none");
  });
});

describe("selectionProblem", () => {
  const covered = [span(NOW - 3 * DAY, NOW)];

  it("passes a well-formed, covered selection", () => {
    expect(selectionProblem({ startMs: NOW - DAY, endMs: NOW - DAY + 60_000 }, covered)).toBeNull();
  });

  it("rejects durations outside the limits", () => {
    expect(selectionProblem({ startMs: NOW - DAY, endMs: NOW - DAY + 1_000 }, covered)).toBe(
      "too-short",
    );
    expect(
      selectionProblem({ startMs: NOW - DAY, endMs: NOW - DAY + MAX_CLIP_MS + 1_000 }, covered),
    ).toBe("too-long");
  });

  it("rejects a range with no footage behind it", () => {
    expect(selectionProblem({ startMs: NOW - DAY, endMs: NOW - DAY + 60_000 }, [])).toBeNull();
    expect(
      selectionProblem({ startMs: NOW - DAY, endMs: NOW - DAY + 60_000 }, [
        span(NOW - 2 * DAY, NOW - 2 * DAY + 1_000),
      ]),
    ).toBe("no-footage");
  });

  it("allows a partially covered range — Frigate returns what exists", () => {
    const sel = { startMs: NOW - DAY, endMs: NOW - DAY + 60_000 };
    expect(selectionProblem(sel, [span(sel.startMs, sel.startMs + 20_000)])).toBeNull();
  });
});

describe("clipRangeSeconds", () => {
  it("rounds outwards so the marked moment is always inside the delivered clip", () => {
    expect(clipRangeSeconds({ startMs: 1_000_400, endMs: 2_000_100 })).toEqual({
      startSec: 1_000,
      endSec: 2_001,
    });
  });

  it("leaves whole seconds untouched", () => {
    expect(clipRangeSeconds({ startMs: 1_000_000, endMs: 2_000_000 })).toEqual({
      startSec: 1_000,
      endSec: 2_000,
    });
  });
});

describe("clipFileName", () => {
  it("stamps the clip's LOCAL start time and its length", () => {
    const start = new Date(2026, 7, 9, 14, 3, 5).getTime();
    expect(clipFileName("big_room", { startMs: start, endMs: start + 50_000 })).toBe(
      "big_room-2026-08-09-14-03-05-50s.mp4",
    );
  });

  it("strips anything that has no business in a filename", () => {
    const start = new Date(2026, 7, 9, 1, 2, 3).getTime();
    const name = clipFileName("../front door", { startMs: start, endMs: start + 10_000 });
    expect(name).toBe("front-door-2026-08-09-01-02-03-10s.mp4");
  });
});
